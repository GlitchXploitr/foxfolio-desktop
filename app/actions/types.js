// @flow

export type Action =
  | { type: 'ADD_SOURCE', source: Object }
  | { type: 'EDIT_SOURCE', source: Object, newSource: Object }
  | { type: 'REQUEST_TRANSACTIONS', source: Object }
  | { type: 'RECEIVE_TRANSACTIONS', exchange: string, trades: Object[], transfers: Object[] }
  | { type: 'FAILED_TRANSACTIONS', exchange: string, error: string }
  | { type: 'TICKER_UPDATE', ticker: Object }
  ;

export type Dispatch = (action: Action | ThunkAction | PromiseAction | Array<Action>) => any;
export type GetState = () => Object;
export type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;
export type PromiseAction = Promise<Action>;
