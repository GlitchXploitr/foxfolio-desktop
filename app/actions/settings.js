// @flow
import type { SettingsType } from '../reducers/settings';
import type { Action } from './action.d';

export function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}
