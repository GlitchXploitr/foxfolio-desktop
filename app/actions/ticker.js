// @flow
import * as R from 'ramda';
import type { Exchanges } from '../reducers/exchanges/types.d';
import type { Action, Dispatch, GetState, ThunkAction } from './action.d';
import startTimer from './timer';
import type { Wallet } from '../reducers/wallets/types.d';

const REFRESH_TIME_IN_MS = 30000;

function fetchingTickerUpdate(): Action {
  return {
    type: 'LAST_UPDATED',
    key: 'ticker',
    time: new Date(),
  };
}

function receiveTickerUpdate(ticker: Object): Action {
  return {
    type: 'TICKER_UPDATE',
    ticker,
  };
}

function receiveHistory(fsym: string, tsym: string, history: [{ close: number }]): Action {
  return {
    type: 'HISTORY_UPDATE',
    fsym,
    tsym,
    history,
  };
}

function receiveCoinList(coinlist: Object): Action {
  return {
    type: 'RECEIVE_COIN_LIST',
    coinlist,
  };
}

export function requestTickerUpdate(
  extraSymbols: string[] = [], fiatCurrency?: string, cryptoCurrency?: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchingTickerUpdate());

    const state = getState();
    const symbols = getSymbolsFromTransactions(
      state.exchanges,
      state.wallets,
      fiatCurrency || state.settings.fiatCurrency,
      cryptoCurrency || state.settings.cryptoCurrency,
      extraSymbols);
    const fsyms = symbols.from.join(',');
    const tsyms = symbols.to.join(',');
    if (fsyms && tsyms) {
      fetch(`https://min-api.cryptocompare.com/data/pricemultifull?fsyms=${fsyms}&tsyms=${tsyms}`)
        .then(result => result.json())
        .then(result => dispatch(receiveTickerUpdate(result.RAW)))
        .catch(error => console.error(error));
    }
  };
}

export function requestHistory(fsym: string, tsym: string, forceRequest: boolean = false): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const twoMinutesAgo = new Date((new Date()).getTime() - (1000 * 60 * 2));
    if (forceRequest
      || !(getState().ticker.history[fsym]
        && getState().ticker.history[fsym][tsym]
        && getState().ticker.history[fsym][tsym].lastUpdate >= twoMinutesAgo)
    ) {
      fetch(`https://min-api.cryptocompare.com/data/histominute?fsym=${fsym}&tsym=${tsym}`)
        .then(result => result.json())
        .then(result => dispatch(receiveHistory(fsym, tsym, result.Data)))
        .catch(error => console.error(error));
    }
  };
}

export function requestCoinList(): ThunkAction {
  return (dispatch: Dispatch) => {
    fetch('https://min-api.cryptocompare.com/data/all/coinlist')
      .then(result => result.json())
      .then(result => dispatch(receiveCoinList(result.Data)))
      .catch(error => console.error(error));
  };
}

export function continuouslyUpdateTicker() {
  return (dispatch: Dispatch, getState: GetState) => {
    if (!getState().timer.ticker) {
      const timer = setInterval(() => dispatch(requestTickerUpdate()), REFRESH_TIME_IN_MS);
      dispatch(startTimer('ticker', timer));
    }
  };
}

function getSymbolsFromTransactions(
  exchanges: Exchanges,
  wallets: Wallet[],
  fiatCurrency: string,
  cryptoCurrency: string,
  extraSymbols: string[]): { from: string[], to: string[] } {
  const walletSymbols = wallets.map(wallet => wallet.currency) || [];
  const exchangeSymbols = R.values(exchanges)
    .map(exchange => exchange.balances)
    .reduce((acc, balances) => acc.concat(R.keys(balances)), []);

  return R.map(R.uniq)(
    {
      from: [cryptoCurrency, ...walletSymbols, ...exchangeSymbols, ...extraSymbols],
      to: [cryptoCurrency, fiatCurrency],
    });
}
