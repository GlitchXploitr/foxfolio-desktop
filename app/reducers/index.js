// @flow
import { combineReducers } from 'redux';
import { routerReducer as router } from 'react-router-redux';
import bittrex from './bittrex';
import counter from './counter';

const rootReducer = combineReducers({
  transactions: bittrex,
  counter,
  router,
});

export default rootReducer;
