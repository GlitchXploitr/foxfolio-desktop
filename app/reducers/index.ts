import { routerReducer as router } from 'react-router-redux';
import { reducer as form } from 'redux-form';
import { persistCombineReducers } from 'redux-persist';
import { configureReduxPersist } from '../store/configureReduxPersist';
import { coinlist, Coinlist } from './coinlist';
import { exchanges } from './exchanges';
import { Exchanges } from './exchanges.types';
import settings, { SettingsType } from './settings';
import { ticker, TickerAndHistory } from './ticker';
import timers, { Timers } from './timer';
import { Wallet, wallets } from './wallets';

const config = configureReduxPersist();
export default persistCombineReducers(config, {
  coinlist,
  exchanges,
  form,
  router,
  settings,
  ticker,
  timers,
  wallets,
});

export interface GlobalState {
  coinlist: Coinlist;
  exchanges: Exchanges;
  settings: SettingsType;
  ticker: TickerAndHistory;
  timers: Timers;
  wallets: Wallet[];
}
