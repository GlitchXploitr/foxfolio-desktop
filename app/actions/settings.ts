import { SettingsType } from '../reducers/settings';
import { Action, Dispatch, ThunkAction } from './actions.types';
import { requestTickerUpdate } from './ticker';

function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}

export function saveSettingsAndUpdateTicker(settings: SettingsType): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(saveSettings(settings));
    dispatch(requestTickerUpdate([], settings.fiatCurrency, settings.cryptoCurrency));
  };
}
