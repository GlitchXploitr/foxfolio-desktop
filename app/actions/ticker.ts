import * as R from 'ramda';
import { Coinlist } from '../reducers/coinlist';
import { Exchanges } from '../reducers/exchanges.types';
import { Ticker } from '../reducers/ticker';
import { Wallet } from '../reducers/wallets';
import { Action, Dispatch, GetState, ThunkAction } from './actions.types';
import startTimer from './timer';

const REFRESH_TIME_IN_MS = 30000;

function fetchingTickerUpdate(): Action {
  return {
    type: 'LAST_UPDATED',
    key: 'ticker',
    time: new Date(),
  };
}

function receiveTickerUpdate(ticker: Ticker): Action {
  return {
    type: 'TICKER_UPDATE',
    ticker,
  };
}

function receiveHistory(fsym: string, tsym: string, history: Array<{ close: number }>): Action {
  return {
    type: 'HISTORY_UPDATE',
    fsym,
    tsym,
    history,
  };
}

function receiveCoinList(coinlist: Coinlist): Action {
  return {
    type: 'RECEIVE_COIN_LIST',
    coinlist,
  };
}

export function requestTickerUpdate(
  extraSymbols: string[] = [],
  fiatCurrency?: string,
  cryptoCurrency?: string
): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchingTickerUpdate());

    const state = getState();
    const symbols = getSymbolsFromTransactions(
      state.exchanges,
      state.wallets,
      fiatCurrency || state.settings.fiatCurrency,
      cryptoCurrency || state.settings.cryptoCurrency,
      extraSymbols
    );
    const fsyms = symbols.from.join(',');
    const tsyms = symbols.to.join(',');
    if (fsyms && tsyms) {
      fetch(`https://min-api.cryptocompare.com/data/pricemultifull?fsyms=${fsyms}&tsyms=${tsyms}`)
        .then(result => result.json())
        .then(result => dispatch(receiveTickerUpdate(result.RAW)))
        .catch(error => console.error(error));
    }
  };
}

export function requestHistory(
  fsym: string,
  tsym: string,
  forceRequest: boolean = false
): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const twoMinutesAgo = new Date(new Date().getTime() - 1000 * 60 * 2);
    if (
      forceRequest ||
      !(
        getState().ticker.history[fsym] &&
        getState().ticker.history[fsym][tsym] &&
        getState().ticker.history[fsym][tsym].lastUpdate >= twoMinutesAgo
      )
    ) {
      fetch(`https://min-api.cryptocompare.com/data/histominute?fsym=${fsym}&tsym=${tsym}`)
        .then(result => result.json())
        .then(result => dispatch(receiveHistory(fsym, tsym, result.Data)))
        .catch(error => console.error(error));
    }
  };
}

export function requestCoinList(): ThunkAction {
  return (dispatch: Dispatch) => {
    fetch('https://min-api.cryptocompare.com/data/all/coinlist')
      .then(result => result.json())
      .then(result => dispatch(receiveCoinList(result.Data)))
      .catch(error => console.error(error));
  };
}

export function continuouslyUpdateTicker(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    if (!getState().timers.timers.ticker) {
      const timer = window.setInterval(() => dispatch(requestTickerUpdate()), REFRESH_TIME_IN_MS);
      dispatch(startTimer('ticker', timer));
    }
  };
}

function getSymbolsFromTransactions(
  exchanges: Exchanges,
  wallets: Wallet[],
  fiatCurrency: string,
  cryptoCurrency: string,
  extraSymbols: string[]
): { from: string[]; to: string[] } {
  const walletSymbols = wallets.map(wallet => wallet.currency) || [];
  const exchangeSymbols = R.pipe(
    R.values,
    R.map(exchange => exchange.balances),
    R.map(R.keys),
    R.reduce((acc, keys) => acc.concat(keys), [] as string[])
  )(exchanges);

  return {
    from: R.uniq([cryptoCurrency, ...walletSymbols, ...exchangeSymbols, ...extraSymbols]),
    to: [cryptoCurrency, fiatCurrency],
  };
}
