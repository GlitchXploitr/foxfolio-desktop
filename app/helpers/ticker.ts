import { Ticker, TickerEntry } from '../reducers/ticker';

export const getTickerEntry = (ticker: Ticker, fsym: string, tsym: string): TickerEntry => {
  if (fsym !== tsym) {
    if (ticker) {
      const tickerBase = ticker[fsym];
      if (tickerBase != null) {
        const tickerEntry = tickerBase[tsym];
        if (tickerEntry != null && tickerEntry.PRICE && tickerEntry.CHANGEPCT24HOUR) {
          return tickerEntry;
        }
      }
    }
    console.log(`Couldn't get ticker entry for ${fsym}/${tsym}`);
    return {
      PRICE: 0,
      CHANGEPCT24HOUR: 0,
    };
  } else {
    return { PRICE: 1, CHANGEPCT24HOUR: 0 };
  }
};
