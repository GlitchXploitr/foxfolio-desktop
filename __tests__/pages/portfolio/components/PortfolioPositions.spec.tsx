import { createShallow } from 'material-ui/test-utils';
import * as React from 'react';
import {
  filterPortfolioForAsset,
  PortfolioPositions,
} from '../../../../app/pages/portfolio/components/PortfolioPositions';
import { Portfolio } from '../../../../app/pages/portfolio/types/portfolio.types';
import { initialSettings } from '../../../../app/reducers/settings';
import { Ticker } from '../../../../app/reducers/ticker';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

test('Render portfolio positions', () => {
  const coinlist = {};
  const portfolio: Portfolio = {
    total: { BTC: 5, ETH: 10 },
    wallets: { BTC: 4, ETH: 8 },
    exchanges: { key: { BTC: 1, ETH: 2 } },
  };
  const ticker: Ticker = {
    BTC: { BTC: { PRICE: 1, CHANGEPCT24HOUR: 0 }, USD: { PRICE: 10000, CHANGEPCT24HOUR: 1.1 } },
    ETH: {
      BTC: { PRICE: 0.1, CHANGEPCT24HOUR: 2 },
      USD: { PRICE: 1000, CHANGEPCT24HOUR: 5.2 },
    },
  };
  const settings = { ...initialSettings, fiatCurrency: 'USD' };

  const wrapper = shallow(
    <PortfolioPositions
      coinlist={coinlist}
      portfolio={portfolio}
      settings={settings}
      ticker={ticker}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('filterPortfolioForAsset', () => {
  const portfolio: Portfolio = {
    total: { BTC: 5, ETH: 10 },
    wallets: { BTC: 4, ETH: 8 },
    exchanges: { key: { BTC: 1, ETH: 2 } },
  };
  const portfolioForAsset = filterPortfolioForAsset(portfolio, 'ETH');

  expect(portfolioForAsset).toMatchObject({ total: 10, wallets: 8, exchanges: { key: 2 } });
});
