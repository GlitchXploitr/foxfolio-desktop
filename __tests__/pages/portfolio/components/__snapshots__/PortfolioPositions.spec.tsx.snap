// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render portfolio positions 1`] = `
<div>
  <WithStyles(Component) />
  <WithStyles(Component)
    asset="BTC"
    coinlist={Object {}}
    key="BTC"
    portfolio={
      Object {
        "exchanges": Object {
          "key": 1,
        },
        "total": 5,
        "wallets": 4,
      }
    }
    settings={
      Object {
        "cryptoCurrency": "BTC",
        "currencyFocus": "fiat",
        "fiatCurrency": "USD",
        "hideZeroBalances": false,
        "includeFiat": true,
        "theme": "light",
      }
    }
    ticker={
      Object {
        "BTC": Object {
          "BTC": Object {
            "CHANGEPCT24HOUR": 0,
            "PRICE": 1,
          },
          "USD": Object {
            "CHANGEPCT24HOUR": 1.1,
            "PRICE": 10000,
          },
        },
        "ETH": Object {
          "BTC": Object {
            "CHANGEPCT24HOUR": 2,
            "PRICE": 0.1,
          },
          "USD": Object {
            "CHANGEPCT24HOUR": 5.2,
            "PRICE": 1000,
          },
        },
      }
    }
  />
  <WithStyles(Component)
    asset="ETH"
    coinlist={Object {}}
    key="ETH"
    portfolio={
      Object {
        "exchanges": Object {
          "key": 2,
        },
        "total": 10,
        "wallets": 8,
      }
    }
    settings={
      Object {
        "cryptoCurrency": "BTC",
        "currencyFocus": "fiat",
        "fiatCurrency": "USD",
        "hideZeroBalances": false,
        "includeFiat": true,
        "theme": "light",
      }
    }
    ticker={
      Object {
        "BTC": Object {
          "BTC": Object {
            "CHANGEPCT24HOUR": 0,
            "PRICE": 1,
          },
          "USD": Object {
            "CHANGEPCT24HOUR": 1.1,
            "PRICE": 10000,
          },
        },
        "ETH": Object {
          "BTC": Object {
            "CHANGEPCT24HOUR": 2,
            "PRICE": 0.1,
          },
          "USD": Object {
            "CHANGEPCT24HOUR": 5.2,
            "PRICE": 1000,
          },
        },
      }
    }
  />
</div>
`;
